build_tools/fish_xgettext.fish

```
@@ -2,84 +2,90 @@
+             string replace -a "\'" "'")
 
 # Extract any constants
 set -a strs (string match -rv 'BUILD_VERSION:|PACKAGE_NAME' -- $expanded |
              string match -rg 'const [A-Z_]*: &str = "(.*)"' | string replace -a "\'" "'")
 
+# Sort the extracted strings and remove duplicates.
+# This is optional.
+set -l strs (for str in $strs
+    echo $str
+end | sort -u)
+
 # We construct messages.pot ourselves instead of forcing this into msgmerge or whatever.
 # The escaping so far works out okay.
 for str in $strs
-    # grep -P needed for string escape to be compatible (PCRE-style),
-    # -H gives the filename.
-    # If you want to run this on non-GNU grep: Don't.
-    # The sed command extracts just the filename from the matches grep finds,
-    # and prepends the '#: ' prefix, marking the line as a source refecence.
-    # sort -u just gets rid of duplicates.
-    grep -PH -r -- \"(string escape --style=regex -- $str)\" src/ |
-       sed -E 's/^([^:]*):.*$/#: \1/' | sort -u
     echo "msgid \"$str\""
     echo 'msgstr ""'
-end >messages.pot
+    echo ""
+end >> $output_file
 
-function extract_fish_script_messages --argument-names name regex;
-    mkdir -p $tmpdir/$name/share/completions $tmpdir/$name/share/functions
-    for f in share/config.fish share/completions/*.fish share/functions/*.fish
-        # Extract messages from fish script.
-        # This is done by creating a file which has the message strings in the same lines as the
-        # oritinal, in the form:
-        # N_ "message"
-        # All other lines will be empty.
-        # Multiple messages on a single line are not supported.
+function extract_fish_script_messages --argument-names regex;
+    # Using xgettext causes more trouble than it helps.
+    # This is due to handling of escaping in fish differing from formats xgettext understands
+    # (e.g. POSIX shell strings).
+    # We work around this issue by manually writing the file content.
```

yeah that's a good approach.
Let's hope msgfmt and friends have good error checking but I don't think the format is likely to change

---

At least when I was still trying to get `xgettext` to work, `msgmerge` complained quite a bit about some things. The dangerous part are quotes which are escaped according to fish syntax but not POSIX syntax. This can result in a large amount of the strings being silently skipped.

---

> quotes which are escaped according to fish syntax but not POSIX syntax.

ugh yeah. I think all our translated strings use fish-style quotes in the source, so `string unescape` is appropriate. Judging from your changes, it looks like the po format only requres escaping backslashes and double quotes.

---

> Judging from your changes, it looks like the po format only requres escaping backslashes and double quotes.

From https://www.gnu.org/software/gettext/manual/html_node/PO-Files.html

> Each of untranslated-string and translated-string respects the C syntax for a character string, including the surrounding quotes and embedded backslashed escape sequences, except that universal character escape sequences (\u and \U) are not allowed.

I have not tested the behavior of various escape sequences extensively, but the current escaping seems to at least ensure that the strings are accepted by msgmerge and msgfmt, without dropping parts of the file due to an incorrectly escaped quote. But more testing would be useful to see if escape sequences work as expected.